### ISSUE TITLE
C11 Atomics: Race condition in atomic_fence

### ISSUE THREAD

**Issue by Jeremy-Kemp** [KhronosGroup/OpenCL-CTS#1274]
We have identified an issue in the atomic_fence test - specifically `release atomic synchronizes-with acquire fence, local atomic_int, scope_work_group` in the case described here.

@kpet pointed out [Internal Bugzilla 15814](https://www.khronos.org/members/login/bugzilla/show_bug.cgi?id=15814) which appears to be the same underlying issue - that instances _may_ make independent forward progress.

The high level summary is that we think that it is possible for the test to fail if interleaving threads execute their instructions in a particular order which is valid, but leads to an incorrect test failure.

Suppose we have 4 instances 0, 1, 2, and 3 which can each make independent forward progress (with a workgroup size of 4).

The following timeline is possible showing how multiple instances accessing the same address races (showing operations from instances 0, 1, and 2):
```C
localValues[myId*1+myValue] = myValue; // Instance 0 writes localValues[1] = 1
atomic_store_explicit(&destMemory[myId], myValue, memory_order_release, memory_scope_work_group); // Instance 0 writes destMemory[0] = 1
hisAtomicValue = atomic_load_explicit(&destMemory[hisId], memory_order_relaxed, memory_scope_work_group); // Instance 0 reads destMemory[3] which is 0 as instance 3 may not have written it yet
localValues[myId*1+myValue] = myValue; // Instance 2 writes localValues[3] = 1.
atomic_work_item_fence(CLK_LOCAL_MEM_FENCE, memory_order_acquire, memory_scope_work_group); // Instance 0
hisValue = localValues[hisId*1+hisAtomicValue];  // Instance 0 Reads localValues[3] which is 1
if(hisAtomicValue != hisValue)     // Instance 0 - At this point hisAtomicValue != hisValue so test failure
```
We do not yet have a proposal for fixing the kernel. Instead, looking to spark some discussion around the issue and test itself so that we can come up with a proposal to fix (whilst maintaining the original intent).

For reference, the entire kernel is:
```C
__kernel void test_atomic_kernel(uint threadCount, uint numDestItems, __global int *finalDest, __global int *oldValues, volatile __local atomic_int *destMemory, __local int *localValues)
{
    uint tid = get_global_id(0);
 
    // initialize atomics not reachable from host (first thread
    // is doing this, other threads are waiting on barrier)
    if (get_local_id(0) == 0)
        for (uint dstItemIdx = 0; dstItemIdx < numDestItems; dstItemIdx++)
        {
            atomic_store_explicit(destMemory + dstItemIdx,
                                  finalDest[dstItemIdx],
                                  memory_order_relaxed,
                                  memory_scope_work_group);
        }
    barrier(CLK_LOCAL_MEM_FENCE); //Redundant
 
    // Copy input reference values into local memory
    localValues[get_local_id(0)] = oldValues[tid];
    barrier(CLK_LOCAL_MEM_FENCE);
 
    size_t myId = get_local_id(0), hisId = get_local_size(0) - 1 - myId;
    int myValue = 0, hisAtomicValue, hisValue;
    do
    {
        myValue++;
        localValues[myId * 1 + myValue] = myValue;
        atomic_store_explicit(&destMemory[myId], myValue, memory_order_release, memory_scope_work_group);
        hisAtomicValue = atomic_load_explicit(&destMemory[hisId], memory_order_relaxed, memory_scope_work_group);
        atomic_work_item_fence(CLK_LOCAL_MEM_FENCE, memory_order_acquire, memory_scope_work_group);
        hisValue = localValues[hisId * 1 + hisAtomicValue];
        hisId = (hisId + 1) % get_local_size(0);
    } while (hisAtomicValue == hisValue && myValue < 1 - 1); // Always false, never loops
    if (hisAtomicValue != hisValue)
    { // fail
        atomic_store(&destMemory[myId], myValue - 1);
        hisId = (hisId + get_local_size(0) - 1) % get_local_size(0);
        if (myValue + 1 < 1)
            localValues[myId * 1 + myValue + 1] = hisId;
        if (myValue + 2 < 1)
            localValues[myId * 1 + myValue + 2] = hisAtomicValue;
        if (myValue + 3 < 1)
            localValues[myId * 1 + myValue + 3] = hisValue;
    }
 
    // Copy local reference values into output array
    barrier(CLK_LOCAL_MEM_FENCE);
    oldValues[tid] = localValues[get_local_id(0)];
 
    // Copy final values to host reachable buffer
    barrier(CLK_LOCAL_MEM_FENCE); //Redundant
    if (get_local_id(0) == 0) // first thread in workgroup
        for (uint dstItemIdx = 0; dstItemIdx < numDestItems; dstItemIdx++)
 
            finalDest[dstItemIdx] =
                atomic_load_explicit(destMemory + dstItemIdx,
                                     memory_order_relaxed,
                                     memory_scope_work_group);
}
```
There is another issue with OoB local memory access, but will directly file a PR for that as its trivial.

**Comment by githubzhu** [comment]
Hi,Jeremy,
Do you have any solution about this issue?
We have the same problem and it happens every time.

**Comment by paulfradgley** [comment]
> Hi, can you please include the test command line that runs this kernel? Thanks!

c11_atomics/test_c11_atomics atomic_fence

**Comment by AhmedAmraniAkdi** [comment]
Hi,

After investigating the issue we found that as Jeremy said, that instances may make independent forward progress because we don't guarantee the synchronisation between threads.

For reference, the part of interest of the kernel is:

```
do
    {
        myValue++;
        localValues[myId * X + myValue] = myValue;
        atomic_store_explicit(&destMemory[myId], myValue, memory_order_release, memory_scope_work_group);
        hisAtomicValue = atomic_load_explicit(&destMemory[hisId], memory_order_relaxed, memory_scope_work_group);
        atomic_work_item_fence(CLK_LOCAL_MEM_FENCE, memory_order_acquire, memory_scope_work_group);
        hisValue = localValues[hisId * X + hisAtomicValue];
        hisId = (hisId + 1) % get_local_size(0);
    } while (hisAtomicValue == hisValue && myValue < X - 1);
```

Where X is returned by NumNonAtomicVariablesPerThread() (c11_atomics/test_atomics.cpp : 1825)

The test may pass even though the threads are reading incorrect values during the do while loop, it just happens that hisAtomicValue == hisValue during the loop and that at the end of the test destMemory[ThreadIdx] == (NumNonAtomicVariablesPerThread() - 1) holds.

Suppose we have 1024 threads, the thread 0 will try to synchronise with thread 1023, the following is valid for when NumNonAtomicVariablesPerThread() is 256:

Suppose thread 1023 is asleep:

thread 0 	

Iteration 1:
myValue = 1
destMemory[0 * 256 + 1] = 1
hisAtomicValue = destMemory[1023] = 0 // thread 1023 hasn't written its value yet.
hisValue = localValues[1023 * 256 + 0] = 0 // thread 1023 hasn't written its value yet.
hisAtomicValue == hisValue is true

Iteration 2:
myValue = 2
destMemory[0 * 256 + 2] = 2
hisAtomicValue = destMemory[1023] = 0 // thread 1023 hasn't written its value yet.
hisValue = localValues[1023 * 256 + 0] = 0 // thread 1023 hasn't written its value yet.
hisAtomicValue == hisValue is true

...
Last iteration
myValue = 255
destMemory[0 * 256 + 255] = 255
hisAtomicValue = destMemory[1023] = 0 // thread 1023 hasn't written its value yet.
hisValue = localValues[1023 * 256 + 0] = 0 // thread 1023 hasn't written its value yet.
hisAtomicValue == hisValue is true

The thread 1023 can wake up and will find hisAtomicValue = destMemory[0] = 255 and hisValue = localValues[0 + 255] = 255 and so hisAtomicValue == hisValue is true, and subsequently destMemory[1023] will be 255 at the end. 

and for the test to pass we need that destMemory[i] to be 255 for all threads which holds. This happens on the lpthreads program that I've attached at the end of this post (same functionality as the OpenCL kernel).

In the situation that Jeremy presented, the test fails at the loop because when the tests assigns only 1 non atomic variable per thread, we can have a situation where we read the element written by another thread that should not participate in the synchronisation, this is not the case for when we have more than 2 non atomic variables per thread.

A solution to this would be to guarantee the synchronisation between threads, according to the [spec](lhttps://en.cppreference.com/w/cpp/atomic/memory_order#Release-Acquire_ordering):

> **Release-Acquire ordering**
> If an atomic store in thread A is tagged memory_order_release and an atomic load in thread B from the same variable is tagged memory_order_acquire, all memory writes (non-atomic and relaxed atomic) that happened-before the atomic store from the point of view of thread A, become visible side-effects in thread B. **That is, once the atomic load is completed, thread B is guaranteed to see everything thread A wrote to memory. This promise only holds if B actually returns the value that A stored, or a value from later in the release sequence.**

The atomic load needs to return the value written by the atomic store in order to ensure synchronisation, a way to do this would be the following:

```
      while(hisAtomicValue < myValue)
      {
           hisAtomicValue = atomic_load_explicit(&destMemory[hisId], memory_order_relaxed);
      }
```

This makes sure that we are reading the correct value and so the threads will advance together.



Another issue is that, for the case of 1 non atomic variable per thread, even though the test works correctly, it will fail because of how verifying the results is done. The condition for the test to pass is:

in c11_atomics/test_atomics.cpp : 2109
```
/* finalValues is populated by the values of destMemory on the OpenCL kernel, so should be last value of myValue. */ HostAtomicType myValue = finalValues[workOffset + i]; 
 ...
/* Which in the case of 1 non atomic variable per thread, we have 1 != 0 and we fail. */
if (myValue != NumNonAtomicVariablesPerThread()-1) 
```
This is mainly because of how the do while loop is set up in the case of 1 non atomic variable per thread.

We can solve this by making it a special case and if NumNonAtomicVariablesPerThread() == 1 then the condition becomes myValue != 1.

**The lpthreads program:**
```
//build with gcc atomic_fence.c -o atomic_fence.out -lpthread

#include <stdlib.h>
#include <pthread.h>
#include <stdatomic.h>
#include <assert.h>
#include <stdio.h>

#define NUM_THREADS 1024
#define MAX_VALUES 256

volatile atomic_int destMemory[NUM_THREADS];
int localValues[NUM_THREADS * MAX_VALUES + 1] = {0};
int output[NUM_THREADS] = {0};
int finalAtomicvalue[NUM_THREADS] = {-1};
int finalDestMemory[NUM_THREADS] = {-1};

pthread_barrier_t barrier;

void *func(void *threadId)
{
    int myId, hisId;
    int hisAtomicValue, hisValue, myValue;

    myId = *((int*) threadId);
    hisId = NUM_THREADS - 1 - myId;
    myValue = 0;
	hisAtomicValue = 0;

    pthread_barrier_wait(&barrier);

    do
    {
        myValue++;
        localValues[myId * MAX_VALUES + myValue] = myValue;
        atomic_store_explicit(&destMemory[myId], myValue, memory_order_release);

        // while(hisAtomicValue < myValue) // needs this loop to work correctly
        // {
            hisAtomicValue = atomic_load_explicit(&destMemory[hisId], memory_order_relaxed);
    	//}
        atomic_thread_fence(memory_order_acquire);
        hisValue = localValues[hisId * MAX_VALUES + hisAtomicValue];

    }while(hisAtomicValue == hisValue && myValue < MAX_VALUES - 1);

    pthread_barrier_wait(&barrier);

    finalAtomicvalue[myId] = hisAtomicValue;

    finalDestMemory[myId] = destMemory[myId];

    if(hisValue != hisAtomicValue)
    {
        output[myId] = -1;
    }
}

int main()
{
    pthread_t threads[NUM_THREADS];
    int thread_args[NUM_THREADS];
    int result = 0;

    for(int i = 0; i < NUM_THREADS; i++)
    {
        atomic_init(&destMemory[i], 0);
    }

    pthread_barrier_init(&barrier, NULL, NUM_THREADS);
    
    for(int i = 0; i < NUM_THREADS; i++)
    {
        thread_args[i] = i;
        result = pthread_create(&threads[i], NULL, func, &thread_args[i]);
        assert(!result);
    }

    for(int i = 0; i < NUM_THREADS; i++)
    {
        result = pthread_join(threads[i], NULL);
        assert(!result);
    }

    pthread_barrier_destroy(&barrier);

    // look for failing tests
    for(int i = 0; i < NUM_THREADS; i++)
    {
        if(output[i] == -1)
        {
            printf("thread %d failed\n", i); // doesn't print anything
        }

	assert(finalDestMemory[i] == (MAX_VALUES - 1)); // never fires
       	assert(finalAtomicvalue[i] == (MAX_VALUES - 1)); // fires
    }
    
    return 0;
}
```

**Comment by b-sumner** [comment]
I think the idea of the test is "If the atomic value from the other thread is not the initial value (so have sync'd) then verify that the fence makes the other thread's normal store visible as well".  I think we just want to fix the code so it does that.  We cannot wait for another lane in any circumstance, we don't have any progress guarantee that allows it.

**Comment by AhmedAmraniAkdi** [comment]
If I understand you correctly, you are saying that:

If hisAtomicValue has changed, then that means the other thread wrote its atomic value, so now the fence guarantees that we see the normal store. But this can happen without the pair of threads advancing together, so the threads won't necessarily read all the values of localValues[threadIdx + hisAtomicValue], they can skip some since hisAtomicValue hasn't changed for that iteration.

The fix I proposed checks that the fence is correct at each iteration of the do while loop, the atomic read inside the while loop guarantees that the pair of threads go together, because it will wait until we get an updated value of hisAtomicValue.

**Comment by b-sumner** [comment]
As I understand/understood the test, there is never a need for the pair to advance together., and that is good because there are not progress guarantees that allow that.  And there is only supposed to be one store to a given location in localValues ever, apart from the initialization.  So I'm not clear on what "all the values of localValues" means - it should be either the initial value, or the single value stored by "him".

**Comment by AhmedAmraniAkdi** [comment]
"all the values of localValues" comes from the fact that you can have more than one non atomic value per thread, this number is returned by NumNonAtomicVariablesPerThread() (c11_atomics/test_atomics.cpp : 1825).

The case Jeremy highlighted has 1 non atomic per thread. And the way the array is set up (if we have x threads, then we have x+1 elements of the localValues array), the failure happens because we may end up reading the element of another thread, thread+1, that doesn't participate in the synchronisation of that pair of threads.

This goes away once you have 2 or more non atomics per thread, because the biggest myValue you can have is 1, you have 2 elements per thread, so you never end up reading the (thread+1) elements.

This can be fixed by having a special case for 1 non atomic per thread, where we will allocate thread_num*2 localValues per thread.

**Comment by b-sumner** [comment]
Umm, OK.  I can't understand why X=1 would ever be a good choice though.
